### DISCUSSION TITLE
Checksum errors in Mac IIci and LCIII

### DISCUSSION THREAD

**Discussion by maciifx** [erichelgeson/BlueSCSI-v1#182] [General]
Hello everyone,

I want to start off by saying that I really love the product and am impressed by its features and build quality.
I do have a rather strange issue that is hard to pinpoint. I have already exhausted the troubleshooting section of the Github page.

First the current setup:
- BlueSCSI with latest firmware (bought fully assembled last week, shipped from Onegeekarmy)
- HP-branded 8GB SD Card (formatted with SD Formatter tool)
- Two HDD-images (created on Linux using dd) formatted with the [patched HD SC Setup](http://macintoshgarden.org/apps/apple-hd-sc-setup-753-patch) (SCSI ID 1 and 2)
- One CD image ([Apple Legacy Recovery CD](http://macintoshgarden.org/apps/apple-legacy-software-recovery-cd)) (SCSI ID 6)
- External SCSI ZIP-drive (added later in the process) (SCSI ID 5)
- Macintosh LCIII (20 MB RAM, PDS network card)
- System 7.5

Past setup:
- Kingston 16GB SD card
- Macintosh IIci (20 MB of RAM, cache card, SuperMac Thunder/24 v1 video card, Apple network card)
- System 7.1.1 (disk images off Legacy Recovery CD)
- System 7.0.0 (physical disks)

The issue:
When I mount disk images from either one of the HDD images or the CD image, most images mount fine but some give me checksum errors. However, when I mount the image again, there is no checksum error. Any and every installation I performed using disk images stored on the virtual media, failed. All installs were done to the virtual disk with SCSI ID 1.
The disk images used come from different sources and at least one set (System 7.5) is verified to be 100% error free.

Steps undertaken:
- Swapped the SD card (no change in behavior)
- Move the BlueSCSI from Macintosh IIci to Macintosh LCIII (no change in behavior)

Observations on the Macintosh IIci:
It boots fine of the virtual CD and virtual HDD (ID 1) with System 7.5 installed. It does not crash and I can reboot it several times without issue. It also boots fine off the Disk Tools floppy containing the patched HD SC Setup.
System 7.5 disk images loaded from physical media (SCSI ZIP-100 drive) don't cause checksum errors and will go through the entire installation process. The same goes for a set of floppies (System 7.0.0).
However, the very same System 7.5 disk images, when copied from the ZIP-disk to the second virtual disk, will start throwing up checksum errors (as stated in the description of the issue, happens randomly to images, not the same ones over and over).
Mounting any set of disk images from the Legacy Recovery CD in the virtual CD drive produces checksum errors in a similar pattern.

Observations on the Macintosh LCIII:
A reproduction of the behavior seen on the Macintosh IIci, including a succesfull installation from disk images mounted directly from the ZIP-100 disk. However, after installation I get an illegal instruction at boot time requiring me to restart the LCIII.

It has mounted up to quite a story. I'm happy to perform any additional troubleshooting steps if required. And just to make it clear: I'm equally convinced there is something wrong with my setup as I am convinced there may be something wrong with the BlueSCSI. No offence meant. Any help is appreciated!

**Comment by erichelgeson** [comment]
Hey sorry you're having trouble, and thanks for the kind words about the project. 

This is an odd one - could you please attach the LOG.txt.

Can you also take a photo or screen shot of the checksum errors?

Lastly - could you try using one of the pre-made blanks in the Usage page under the first link. Those are made by DiskJockey and will give us a common base to test with.

**Comment by maciifx** [comment]
I'll provide the data in the order you requested it, starting with the screenshots. The first screenshot shows the checksum error Disk Copy throws when mounting the image. I mounted it again immediately after and it is fine (second screenshot).

![IMG_20221012_102400464](https://user-images.githubusercontent.com/115500465/195299923-4685f072-ec59-464d-ad94-fbe4ed7390dd.jpg)
![IMG_20221012_102431319](https://user-images.githubusercontent.com/115500465/195299920-caedffc4-386b-483e-95f7-c9aed1ce7516.jpg)
These were made with the setup used in the opening post. Here is the corresponding log file:
[LOG-20221012-1031.txt](https://github.com/erichelgeson/BlueSCSI/files/9763348/LOG-20221012-1031.txt)

I have tested two setups, while I was at it. 

Setup 1:
I downloaded [a premade image from Macintosh Garden](http://macintoshgarden.org/apps/rascsi-reloaded) of System 7.5.3 and placed it on the SD card. I duplicated it and gave it a different SCSI ID so I had two SCSI disks. Left the first one as it was, inserted the SD card in the LCIII and booted it (which worked fine). I used the patched HD SC Setup to initialize the second HDD and copied Disk Copy and disk images over from the ZIP-disk. Ordered Disk Copy to mount all disk images but the error remains. If anything, the failure rate seems to have increased, because only half the images would mount (before it would be one or two).

Corresponding log file: [LOG-20221012-1050.txt](https://github.com/erichelgeson/BlueSCSI/files/9763347/LOG-20221012-1050.txt)

Setup 2:
I downloaded the [premade blank disk images](https://mega.nz/folder/8hA3AQCJ#pWUq92L70yDXlogy9lk5Dg/folder/01hkVLoC) from the BlueSCSI Wiki (the 100 MB version). I duplicated it just like in setup 1 and additionally removed all other images from the SD card.
I made no further changes, placed the SD card in the LCIII, booted off the ZIP-disk, copied over Disk Copy and the disk image files and mounted them. Results are on par with setup 1, mounting approximately half the disk images.

Corresponding log file:[LOG-20221012-1102.txt](https://github.com/erichelgeson/BlueSCSI/files/9763346/LOG-20221012-1102.txt)
Corresponding screenshot (of the last image that had a checksum error):
![IMG_20221012_110054561](https://user-images.githubusercontent.com/115500465/195299908-b45b7445-d159-4233-9586-576beae9f0f9.jpg)

I stuck with the LCIII for testing, based on the fact it produced similar results with similar symptoms as did the Mac IIci.
If need be, I can retry the test with a Mac IIfx, Quadra 700 or Power Mac 6100.

Possibly also of importance: any file handling on the SD card for these setups was done using Linux (specifically Kubuntu 22.04).
I have also used Windows 7 to copy files onto the SD card in the opening post, after formatting it with the SD Card Formatter.

Let me know what you think.

**Comment by maciifx** [comment]
Hello Eric,

Just to let you know I tried a third known-brand SD-card (SanDisk this time) with just a blank 100MB image, but the results are the same as before. I think it to be highly unlikely that the problem is related to any individual SD-card.

**Comment by maciifx** [comment]
Well, this evening I received an e-mail from Jean-Michel over at Onegeekarmy, stating the following:

_You have recently purchased a BlueSCSI from me and I wanted to inform you about a potential issue I’ve discovered, as well as a way to solve it.

As I mention in the little note that accompanies the shipment, your STM32 runs faster than the usual 72 MHz. It turns out that the STM32 I have received from my supplier are actually not STM32s but APM32s.
As you’ll notice, the chip is not labeled as such, which leads to further confusion. Both STM32 and SPM32 work almost identically, but not exactly.

In essence, the APM32 has a larger memory pool (256 KB instead of the 128 KB of the STM32) and it’s a little faster for a few operations. For its use in BlueSCSI, a faster version of the firmware is needed. This is why yours was flashed with the 96 MHz firmware. As it turns out, in some cases, this is not enough._

He has provided a link to the most recent firmware:
https://github.com/erichelgeson/BlueSCSI/releases/tag/v1.1-20220917

Hoping this might resolve the issues I have seen, I updated the firmware on my BlueSCSI, which completed succesfully.
Next I reran the entire process:
- Formatted SD card with SD Card Formatter
- Formatted in Windows to exFat
- copied over the empty 100MB virtual disk
- hooked up the Mac IIci and Zip-drive
- copied the System 7.5 disk images to the empty virtual disk
- mounted the disk images in Disk Copy 6.3.3

Short version: it does not solve the problem.
Long version: I see different symptoms now. When mounting the first image, the progress bar now halts with the HDD activity LED on full. Disk Copy then tells me something is wrong with the disk image, aborts mounting the image and proceeds to the next one. This happens again further down the road. In addition to that, Disk Copy now reports checksum errors on ALL disk images (apart from the ones it aborted mounting). 

Things I have tried:
- Reflashing the BlueSCSI with the latest (128MHz) firmware
- Switched SD-cards

I have attached the log file from the SD card.
[LOG-20221023-2110.txt](https://github.com/erichelgeson/BlueSCSI/files/9847333/LOG-20221023-2110.txt)

**Comment by maciifx** [comment]
Any thoughts on this?
